web_poet/testing/pytest.py

```diff
@@ -0,0 +1,90 @@
+import asyncio
+import json
+from pathlib import Path
+from typing import Any, Iterable, Optional, Set, Union
+
+import pytest
+
+from web_poet import ItemPage
+from web_poet.serialization import SerializedDataFileStorage, deserialize, load_class
+from web_poet.testing import INPUT_DIR_NAME, OUTPUT_FILE_NAME
+from web_poet.utils import ensure_awaitable
+
+
+class WebPoetFile(pytest.File):
+    """Represents a directory containing test case subdirectories for one page object."""
+
+    def collect(self) -> Iterable[Union[pytest.Item, pytest.Collector]]:  # noqa: D102
+        for entry in self.path.iterdir():
+            if entry.is_dir():
+                item: WebPoetItem = WebPoetItem.from_parent(
+                    self,
+                    name=entry.name,
+                    path=entry,
+                )
+                if item.input_path.is_dir() and item.output_path.is_file():
+                    # item.add_marker(pytest.mark.asyncio)
+                    yield item
+
+
+class WebPoetItem(pytest.Item):
+    """Represents a directory containing one test case."""
+
+    @property
+    def po_name(self) -> str:
+        """The type name of the PO being tested."""
+        if not self.parent:
+            raise ValueError("WebPoetItem has no parent")
+        return self.parent.name
+
+    @property
+    def input_path(self) -> Path:
+        """The inputs subdirectory path."""
+        return self.path / INPUT_DIR_NAME
+
+    @property
+    def output_path(self) -> Path:
+        """The output file path."""
+        return self.path / OUTPUT_FILE_NAME
+
+    def get_page(self) -> ItemPage:
+        """Return the page object created from the saved input."""
+        po_type = load_class(self.po_name)
+        if not issubclass(po_type, ItemPage):
+            raise TypeError(f"{self.po_name} is not a descendant of ItemPage")
+        storage = SerializedDataFileStorage(self.input_path)
+        return deserialize(po_type, storage.read())
+
+    async def get_po_output(self) -> dict:
+        """Return the output from the PO."""
+        po = self.get_page()
+        return await ensure_awaitable(po.to_item())
+
+    def get_expected_output(self) -> dict:
+        """Return the saved output."""
+        return json.loads(self.output_path.read_bytes())
+
+    def runtest(self) -> None:  # noqa: D102
+        output = asyncio.run(self.get_po_output())
+        expected_output = self.get_expected_output()
+        assert output == expected_output
+
+
+_found_po_dirs: Set[Path] = set()
+
+
+def pytest_collect_file(
+    file_path: Path, path: Any, parent: pytest.Collector
+) -> Optional[pytest.Collector]:
+    if file_path.name == OUTPUT_FILE_NAME:
```

Comment:
output.json is quite a common name; do you see any issues with that?

Reply:
Yeah, but as far as I can see pytest collection works on files and there seems to be no better way to identify a test case. We could add more checks here, e.g. for the inputs dir. Or we could use some prefix for output.json, a potential meta.json etc. and move deps into the same dir, so that ignoring them when readins serialized deps is easier and seeing them when looking at the dir content manually is hopefully easier too.

Reply:
> Sorry, I didn't quite catch this. Could you please elaborate?

This is related to the earlier discussion in this PR: we could put output.json and the serialized deps in the same directory and I suggested we could add some other files there which are not serialized deps. All those files could have some common prefix.